### PR TITLE
⚡️ add speedy mode for sandbox to improve performance

### LINKED ISSUE #588: [Bug] vue 项目中，在页面 checkbox 比较多的情况下，切换 checkbox 状态导致卡顿
## What happens?

在 vue 项目中，当页面 checkbox 比较多的情况下，改变 checkbox 状态会导致页面卡顿，element 和 iview 的 checkbox 都可以复现

## 最小可复现仓库

https://github.com/everyun/qiankun-vue-demo

### 运行
master 为基座，module 为微应用  
分别 npm install && npm run serve 之后  
主项目访问地址 http://localhost:10010/simple/  
微应用直接地址为 http://localhost:10011/simple/  

### 复现
页面中放了 200 个 checkbox，当切换 checkbox 状态时，可以感觉到明显的卡顿  
直接访问微应用做同样的操作，没有卡顿的现象

### LINKED ISSUE #1110: [Bug]qiankun + vue3 + antd2 select组件卡顿
## What happens?

qiankun + vue3 + antd2 select组件卡顿(虚拟列表)，单独运行流畅，接入qiankun卡顿。
table组件在切换pageSize时( 30 -> 200 -> 30)有同样的卡顿问题。

## Mini Showcase Repository(REQUIRED)

https://github.com/xiaoLus/qiankun-vue3-antd2

### 运行

main 为基座，subapp 为微应用  
分别 yarn && yarn serve 之后  
主项目访问地址 <http://localhost:10010/>  
微应用直接地址为 <http://localhost:10011/>

### 复现

使用 ant-design-vue@2.0.0 官方文档 select组件的 big data 功能。  
微应用单独启动时非常流畅，  
嵌入 qiankun基座 中非常卡。  
<https://2x.antdv.com/components/select-cn/#components-select-demo-big-data>

**Expected behavior** 
select组件滚动流畅无卡顿

## Context

- **qiankun Version**: 2.3.2
- **Platform Version**: win10 & macOS catalina
- **Browser Version**: chrome 87

### LINKED ISSUE #1175: [Feature Request] 在沙盒里面对 document 也做代理会不会更好？
## Background

大概从 qiankun@2.1.0 以后，以 proxySandbox 作为默认的沙盒，并且默认开启 strickGlobal 特性，基于 with 语句执行子应用脚本。我们的测试结果显示，在这种模式下，单纯的脚本执行性能还可以，但是一旦涉及到操作 DOM 时，性能就会严重下降，大概会差 10 倍以上。

## Mini Showcase Repository

[https://gitee.com/kawhi66/test/tree/master/qiankun@2.3.4](https://gitee.com/kawhi66/test/tree/master/qiankun@2.3.4)

## Proposal

一个想法，既然沙盒可以提供独立的 window 实例，是否也应该提供独立的 document 实例 ？

## Additional context

我们做了简单的验证，在 import-html-entry 里面做如下的修改：

```js
// src/index.js
function getExecutableScript(scriptSrc, scriptText, proxy, strictGlobal) {
	const sourceUrl = isInlineCode(scriptSrc) ? '' : `//# sourceURL=${scriptSrc}\n`;

	// 通过这种方式获取全局 window，因为 script 也是在全局作用域下运行的，所以我们通过 window.proxy 绑定时也必须确保绑定到全局 window 上
	// 否则在嵌套场景下， window.proxy 设置的是内层应用的 window，而代码其实是在全局作用域运行的，会导致闭包里的 window.proxy 取的是最外层的微应用的 proxy
	const globalWindow = (0, eval)('window');
	globalWindow.proxy = proxy;
	// TODO 通过 strictGlobal 方式切换切换 with 闭包，待 with 方式坑趟平后再合并
	return strictGlobal
                // 修改的地方在这里，with 语句将 document 也加进去
		? `;(function(window, self, document){with(window, document){;${scriptText}\n${sourceUrl}}}).bind(window.proxy)(window.proxy, window.proxy, window.document);`
		: `;(function(window, self){;${scriptText}\n${sourceUrl}}).bind(window.proxy)(window.proxy, window.proxy);`;
}
```

测试结果显示，在操作 DOM 时，性能会提升不少，大概会好 60%，虽然还是会慢，但会好很多。

### LINKED ISSUE #1223: [Feature request]执行代码过慢
umi这个模块执行大概要1s，有什么方式可以去优化的呢？拆包吗？？
image

### LINKED ISSUE #1320: 接入qiankun 框架后,卡顿现象凸显的问题
## 现状

1. 微应用独立运行时，有轻微卡顿现象
2. 引入qiankun 框架后，卡顿现象明显，性能问题凸显

## 最小复现仓库

qiankun-better 问题demo

#### 访问项目
- 在qiankun 框架内访问卡顿页面： http://localhost:8080/#/data  
- 独立访问微应用卡顿页面：http://localhost:8082/#/data
- 
#### 操作步骤
- 点击表头　“操作”　列，弹出复选框，控制表格列的动态显示

#### 卡顿表现
- 独立访问微应用，表格列的显示过程有轻微卡顿
- qiankun 框架访问, 表格列的显示过程卡顿明显, 延迟时间长
** (1. 直观操作感觉, 2 .通过 devtools  performence 性能分析,  summary  统计时长得出的)**

## 环境

- **qiankun Version**: 2.3.5 及以上
- **Platform Version**:  nodejs 环境 10.13.0 (nodejs 版本在 8.11 以上即可)
- **Browser Version**: chrome 88 ( 对chrome 版本没要求, 常用版本即可)

### LINKED ISSUE #1555: [RFC]  qiankun的极致性能优化思路，与import-html-entry有关
## 背景
大概从 qiankun@2.1.0 以后，以 proxySandbox 作为默认的沙盒，并且默认开启 strickGlobal 特性，基于 with 语句执行子应用脚本。我们的测试结果显示，在这种模式下，单纯的脚本执行性能还可以，但是一旦涉及到操作 DOM 时，性能就会严重下降，大概会差 10 倍以上。

> 发现问题和提供解决思路的issues，要特别鸣谢哦！
[[Feature Request] 在沙盒里面对 document 也做代理会不会更好？ #1175](https://github.com/umijs/qiankun/issues/1175)

当前我们的项目中使用react+antd技术栈，由于是B端项目，存在大量的antd组件，而antd组件会生成大量的dom节点，从而导致渲染慢，性能很差。

**可以发现性能损失主要来源于以下两点：**
1. 相对于沙盒带来的性能损失远比带来的好处要小的多，故可忽略沙盒的性能问题。
2. 但基于import-html-entry的with执行子应用脚本来说，这里的性能下降及其严重。
3. 当子应用懒加载一个大的模块时，如果涉及到dom操作会导致明细卡顿，fps直接为0了

## qiankun 版本
以qiankun@2.1.0 以后为主，之前的版本没测试，应该也能使用，但思路都是一致的。

## 最小可复现仓库
[可使用#1175提供的仓库 https://gitee.com/kawhi66/test/tree/master/qiankun@2.3.4](https://gitee.com/kawhi66/test/tree/master/qiankun@2.3.4)

一个想法，既然沙盒可以提供独立的 window 实例，是否也应该提供独立的 document 实例 ？

- qiankun默认的性能
image
- 沙盒再提供一个独立的document实例，此时的性能
将import-html-entry的getExecutableScript方法中的`with(window)`改成`with(window, window.document)`
```js
// src/index.js
function getExecutableScript(scriptSrc, scriptText, proxy, strictGlobal) {
	const sourceUrl = isInlineCode(scriptSrc) ? '' : `//# sourceURL=${scriptSrc}\n`;

	// 通过这种方式获取全局 window，因为 script 也是在全局作用域下运行的，所以我们通过 window.proxy 绑定时也必须确保绑定到全局 window 上
	// 否则在嵌套场景下， window.proxy 设置的是内层应用的 window，而代码其实是在全局作用域运行的，会导致闭包里的 window.proxy 取的是最外层的微应用的 proxy
	const globalWindow = (0, eval)('window');
	globalWindow.proxy = proxy;
	// TODO 通过 strictGlobal 方式切换切换 with 闭包，待 with 方式坑趟平后再合并
	return strictGlobal
                // 修改的地方在这里，with 语句将 document 也加进去
		? `;(function(window, self){with(window, window.document){;${scriptText}\n${sourceUrl}}}).bind(window.proxy)(window.proxy, window.proxy);`
		: `;(function(window, self){;${scriptText}\n${sourceUrl}}).bind(window.proxy)(window.proxy, window.proxy);`;
}
```
image

> 由此可见性能提升3.5-4倍

但是官方不使用此方法的原因主要是：
1. 疑似导致document无法被沙盒代理；
2. with的语法特性导致的严重bug；
> 测试URL和titile属性：URL在window中是一个方法，但在document中是链接
image
> 由此可见with代理document会产生严重的问题，不过我们通过规范代码来尽量规避这些问题。既然如此就赶紧在项目中使用，毕竟解决了性能问题，可是实际使用发现了另一个重大bug：无法使用懒加载。。。
image
产生这个原因：webpack懒加载是用过document.head.appendChild(script)来解决的。
with代理的document会导致懒加载失效，所以我们需要转变思路。

3. 解决with代理document产生的问题的思路
```js
/* 沙盒注入import-html-entry的execScripts方法 */
// 1. 通过qiankun的registerMicroApps的beforeLoad方法的第二个参数来注入execScripts方法
// 我就简单点，直接在beforeLoad中注入。
// qiankun团队可beforeLoad调用之前就注入execScripts
import { execScripts } from 'import-html-entry';
registerMicroApps([
  {
    name: "vue app",
    entry: "http://localhost:5101",
    container: "#subapp-container",
    activeRule: "/vue"
  }], {
    beforeLoad: function(app, global) {
      global.execScripts = execScripts;
    }
  }
);
// 2.修改import-html-entry的getExecutableScript方法，通过重写document.head.appendChild方法来实现懒加载
function getExecutableScript(scriptSrc, scriptText, proxy, strictGlobal) {
  const sourceUrl = isInlineCode(scriptSrc) ? '' : `//# sourceURL=${scriptSrc}\n`;
  const globalWindow = (0, eval)('window');
  globalWindow.proxy = proxy;
  const execScripts = `
    var headAppendChild = document.head.appendChild.bind(document.head);
    document.head.appendChild = function(elem) {
      if(elem.nodeName.toLocaleLowerCase() === 'script') {
        /*execScripts参数解析
          execScripts({
            entry,
            scripts,
            proxy,
            opts:{
              fetch：传入fetch或使用defaultFetch（即window.fetch.bind(window)）
              strictGlobal：默认false，
              success：成功的回调函数
              error：失败的回调函数
              beforeExec：js代码字符串前置执行函数
              afterExec：js代码字符串后置执行函数
            }
          })
        */
        (typeof self !== 'undefined' ? self : this)['execScripts'](
          elem.src,
          [elem.src],
          window, // proxy
          {
            fetch: window.fetch,
            strictGlobal: true,
            success: elem.onload,
            error: elem.error
          }
        );
      } else {
        headAppendChild(elem);
      }
    };
  `;
  // TODO 通过 strictGlobal 方式切换切换 with 闭包，待 with 方式坑趟平后再合并
  return strictGlobal
    ? `;(function(window, self){with(window, window.document){;${execScripts}\n${scriptText}\n${sourceUrl}}}).bind(window.proxy)(window.proxy, window.proxy);`
    : `;(function(window, self){;${scriptText}\n${sourceUrl}}).bind(window.proxy)(window.proxy, window.proxy);`;
}
```
> 整合上面的思路
* 第一：沙盒注入import-html-entry的execScripts；
* 第二：with注入document；`with(window, document)`
* 第三：重写import-html-entry的getExecutableScript，代理子应用js的document.head.appendChild方法，并使用注入的execScripts来解析子应用的js
* 总结：解决了上面的懒加载的问题，性能与直接在with中注入document一致

4. 又可以愉快的使用高性能的qiankun啦。但是仍然没有解决`with(window, document)`的属性冲突问题。。。
> `with(window, document)`会导致window.URL(方法)被document.URL(链接)覆盖，导致无法使用`new URL()`了
> 反过来`with(document, window)`会导致document.URL(链接)被window.URL(方法)覆盖，导致无法使用`document.URL`了
> 不过规范代码，也是可以继续写项目。
```js
/*
研究了一下with语法，发现
obj, obj1, ...., objn 相当于Object.assign({}, obj, obj1, ..., objn);
由此可见with(window, document)会导致window.URL被document.URL覆盖
*/
with(obj, obj1, ..., objn) {
   console.log(key)
}
```

5. 想了想重写的getExecutableScript方法：最后return这段
`;(function(window, self){with(window, window.document){;${execScripts}\n${scriptText}\n${sourceUrl}}}).bind(window.proxy)(window.proxy, window.proxy);`
> 看这段代码，就知道已通过bind方法将函数的this修改为沙盒proxy，回调函数也已经注入了window和self。
> 既然我重写了getExecutableScript方法，通过注入的execScripts解析了子应用的js，那么又何必使用使用with呢
> 在不使用with的情况下，就没有了with产生的bug了。

**qiankun极致性能优化：**将重写的getExecutableScript方法中的with(window, window.document){}删除
> 极致性能，测试结果如下，而且与不使用qiankun的子应用的性能相差无几。
image

### 总结
1.  beforeLoad注入import-html-entry的execScripts方法，可交由qiankun的loader.js来实现，在生成沙盒之后注入execScripts方法
2. 重写import-html-entry的getExecutableScript方法应该可以交由beforeExec和afterExec来实现，或者qiankun直接重写该方法
3. 子应用重写document.head.appendChild方法可webpack plugin来实现重写requireEnsure方法，根据是否qiankun子应用使用execScripts或document.head.appendChild
4. 尚不知此番操作会不会导致沙盒的内存溢出问题 或者 其他隐性bug问题，所以以上操作目前来说最好在个人项目，内部项目使用。
5. 第一次优化提升3.5-4倍性能，第二次优化提升16倍性能。
6. 希望qiankun团队可以采用本方案，或者以此方案想出更好的性能优化方案。

### LINKED ISSUE #1944: qiankun 微应用使用 echarts 地图钻取时有明显卡顿现象
## 现状?

1.微应用独立运行时echarts地图钻取流畅
2.微应用嵌入qiankun，地图钻取时，有明显的卡顿
3.关闭 sandbox 就不会出现 2 描述的现象

## 最小可复现仓库

https://github.com/yl196356303/qinkun-echarts-demo

## 复现步骤，错误日志以及相关配置

1.qiankun加载子应用 http://localhost:8080/micro-app/
2.钻取地图出现明显卡顿
3.http://localhost:7071/ 独立运行微应用钻取地图很流畅
4.关闭 sandbox 就不会出现

## 相关环境信息

- **qiankun 版本**:  2.5.1
- **浏览器版本**：chrome  97.0.4692.71
- **操作系统**：windows 10
-**node版本**：v14.15.0

### LINKED ISSUE #1980: 主应用和子应用都用angular12+ zorro, 在某些下拉框多的模态框打开后，页面非常的卡顿
复现条件：
主应用和子应用都用angular12+ zorro, 
在某些下拉框多的模态框打开后，
页面非常的卡顿

第一：卡顿的问题是否可以解决？
第二：有一个不完美的解决方案：关闭sandbox后性能问题解决，但是关闭sandbox后导航无法跳转，浏览器前进后退也失效？

### LINKED ISSUE #2008: [BUG]子应用如果带有 web worker，在接入到子应用后会被阻塞延迟一段时间才会被调用。
## 子应用如果带有 web worker，在接入到子应用后会被阻塞延迟一段时间才会被调用。
我们有个子应用是使用 Cesium 进行三维数据的渲染，当我们使用 Cesium 去渲染 primitve 时候，独立测试都很正常，而在接入主运用后 Cesium 渲染 primitve  的 web worker 会被阻塞10几秒。

## 相关环境信息
- qiankun 版本： 2.6.3

### LINKED ISSUE #2035: qiankun下大3D文件渲染很慢
## What happens?

我将一个3D文件的查看界面 https://github.com/kovacsv/Online3DViewer 变成了qiankun的 子项目，他是一个html项目，用nginx跑起来后，发现qiankun下大3D文件的渲染速度比单独启动慢一分钟

## Mini Showcase Repository(REQUIRED)

https://github.com/kovacsv/Online3DViewer

## How To Reproduce

https://github.com/kovacsv/Online3DViewer         子应用

## Context

- **qiankun Version**:
- **Platform Version**:
- **Browser Version**:

### PR DESCRIPTION
**Pull request by kuitos**
Use `sandbox: { speedy: true }` to enable this performance improvement.

ref https://github.com/kuitos/import-html-entry/pull/82

resolves #2035 
resolves #2008 
resolves #1980 
resolves #1944 
resolves #1555 
resolves #1320 
resolves #1223 
resolves #1175 
resolves #1110 
resolves #588

### CODE REVIEW

File: src/sandbox/common.ts [hunk 1]
```diff
@@ -77,15 +79,29 @@ export function getTargetValue(target: any, value: any): any {
...
+    'Reflect',
+    'Float64Array',
+    'Function',
+    'Map',
+    'NaN',
+    'Promise',
+    'Proxy',
+    'Set',
+    'parseInt',
+    'requestAnimationFrame',
```
Comment (fengmk2) [inline src/sandbox/common.ts:105]: 后续每新增一个 window 全局变量都可以来这里蹭贡献了。

Comment (kuitos) [reply src/sandbox/common.ts:105]: 哈哈哈机智

Comment (kuitos) [reply src/sandbox/common.ts:105]: 这里其实可以换成黑名单机制，默认把原生 window 上的全局变量遍历出来，只剔除黑名单里 qiankun 需要额外关注的全局变量，确保其还是走沙箱的 getter/setter

Comment (hstarorg) [reply src/sandbox/common.ts:105]: 黑名单的话，如果子应用在覆写某些原型链方法会不会拦不住？

Comment (kuitos) [reply src/sandbox/common.ts:105]: 原型链一直都没拦的

File: src/sandbox/common.ts [hunk 2]
```diff
@@ -77,15 +79,29 @@ export function getTargetValue(target: any, value: any): any {
...
+  'Reflect',
+  'Float64Array',
+  'Function',
+  'Map',
+  'NaN',
+  'Promise',
+  'Proxy',
+  'Set',
+  'parseInt',
+  'requestAnimationFrame',
```
Comment (afc163) [inline src/sandbox/common.ts:104]: 能直接用 https://github.com/sindresorhus/globals/blob/main/globals.json 么

Comment (kuitos) [reply src/sandbox/common.ts:104]: 看上去可以，不过这是个 json，直接 import 的话依赖 bundler 的配置（比如 webpack），没有配置 json loader 的应用 import qiankun 可能会报错。不过可以在 qiankun build 阶段预处理一下来解决这个问题。

Comment (kuitos) [reply src/sandbox/common.ts:104]: builtin 的可以用，browser 的还不行，有些 bom 变量不能被 unscoped，比如 parent

Comment (kuitos) [reply src/sandbox/common.ts:104]: https://github.com/umijs/qiankun/pull/2288

**Comment by shipMatserJack** [comment]
nice

**Comment by CookyG** [comment]
mark

**Pull request merged**